src/libopenrave/trajectory.cpp

```diff
@@ -21,54 +21,194 @@
+
+    /* Store meta-data */
+
+    // Indicate size of meta data
+    const ConfigurationSpecification& spec = this->GetConfigurationSpecification();
+    const uint16_t numGroups = spec._vgroups.size();
+    WriteBinaryUInt16(O, numGroups);
+
+    for (uint16_t groupIndex = 0; groupIndex < numGroups; groupIndex++)
+    {
+        const ConfigurationSpecification::Group& group = spec._vgroups[groupIndex];
+        WriteBinaryString(O, group.name);   // Writes group name
+        WriteBinaryInt(O, group.offset);    // Writes offset
+        WriteBinaryInt(O, group.dof);       // Writes dof
+        WriteBinaryString(O, group.interpolation);  // Writes interpolation
     }
-    O << "</trajectory>" << endl;
+
+    /* Store data waypoints */
+    std::vector<dReal> trajectoryData;
+    this->GetWaypoints(0, this->GetNumWaypoints(), trajectoryData);
+    WriteBinaryVector(O, trajectoryData);
 }
 
 InterfaceBasePtr TrajectoryBase::deserialize(std::istream& I)
 {
-    stringbuf buf;
-    stringstream::streampos pos = I.tellg();
-    I.get(buf, 0); // get all the data, yes this is inefficient, not sure if there anyway to search in streams
-    BOOST_ASSERT(!!I);
-
-    string pbuf = buf.str();
-    const char* p = strcasestr(pbuf.c_str(), "</trajectory>");
-    int ppsize=-1;
-    if( p != NULL ) {
-        I.clear();
-        ppsize=(p-pbuf.c_str())+13;
-        I.seekg((size_t)pos+ppsize);
+    // Check whether binary or XML file
+    I.seekg(0, I.beg);
+    uint16_t binaryFileHeader = 0;
+    ReadBinaryUInt16(I, binaryFileHeader);
+
+    // Read binary trajectory files
+    if (binaryFileHeader == MAGIC_NUMBER)
+    {   
+        uint16_t versionNumber = 0;
+        ReadBinaryUInt16(I, versionNumber);
+
+        if (versionNumber != VERSION_NUMBER)
+        {
+            // Throw assertion (for future versions)
+        }
+            
+        /* Read metadata */
+
+        // Read number of groups
+        uint16_t numGroups = 0;
+        ReadBinaryUInt16(I, numGroups);
+
+        ConfigurationSpecification spec;
+        spec._vgroups.resize(numGroups);
+        for (unsigned int groupIndex = 0; groupIndex < numGroups; groupIndex++)
+        {
+            ReadBinaryString(I, spec._vgroups[groupIndex].name);    // Read group name
+            ReadBinaryInt(I, spec._vgroups[groupIndex].offset);     // Read offset
+            ReadBinaryInt(I, spec._vgroups[groupIndex].dof);        // Read dof
+            ReadBinaryString(I, spec._vgroups[groupIndex].interpolation);   // Read interpolation
+        }
+        this->Init(spec);
+
+        /* Read trajectory data */
+        std::vector<dReal> trajectoryData;
+        ReadBinaryVector(I, trajectoryData);
+        this->Insert(this->GetNumWaypoints(), trajectoryData);
     }
-    else {
-        throw OPENRAVE_EXCEPTION_FORMAT(_("error, failed to find </trajectory> in %s"),buf.str(),ORE_InvalidArguments);
+    // Backwards compatible with old XML trajectory files
```

Comment:
I suggest you put a `return` before the `else` and keep the legacy part of code untouched with no indentation.